### PR TITLE
868 msa precip distribution bimodality

### PR DESCRIPTION
**Pull request by msahn**

### CODE REVIEW

**Comment by lee1043** [comment]
@msahn I have conducted `pre-commit` test and applied `autoflake8` and `black` which automatically adjust code styling to follow PEP8, and also manually revised several lines. Some changes still need to be made, and as they may related the way code works, I am leaving them to you. Please check below text file that logged pre-commit test output to check which lines need to be revised. Please let me know if you have any question.

[pre-commit-log.txt](https://github.com/PCMDI/pcmdi_metrics/files/11176794/pre-commit-log.txt)

For the conflicts, I recommend checking in manual, as some of them may have been affected by the recent changes.

**Comment by msahn** [comment]
> @msahn I have conducted `pre-commit` test and applied `autoflake8` and `black` which automatically adjust code styling to follow PEP8, and also manually revised several lines. Some changes still need to be made, and as they may related the way code works, I am leaving them to you. Please check below text file that logged pre-commit test output to check which lines need to be revised. Please let me know if you have any question.
> 
> [pre-commit-log.txt](https://github.com/PCMDI/pcmdi_metrics/files/11176794/pre-commit-log.txt)
> 
> For the conflicts, I recommend checking in manual, as some of them may have been affected by the recent changes.

Thank you for your help! I have manually updated the code based on [pre-commit-log.txt](https://github.com/PCMDI/pcmdi_metrics/files/11176794/pre-commit-log.txt). I will also check which parts are conflicting with main branch.

**Comment by lee1043** [comment]
@msahn thanks, I did pre-commit test based on the latest in this branch. Only one more to revise. Nice work!

```
pcmdi_metrics/precip_distribution/lib/lib_precip_distribution.py:552:5: F841 local variable 'wm2tommd' is assigned to but never used
```

**Comment by msahn** [comment]
> @msahn thanks, I did pre-commit test based on the latest in this branch. Only one more to revise. Nice work!
> 
> ```
> pcmdi_metrics/precip_distribution/lib/lib_precip_distribution.py:552:5: F841 local variable 'wm2tommd' is assigned to but never used
> ```

Thank you. I have fixed it.
Also, I have manually resolved the conflicts between this branch and main branch. I will test the code.

**Comment by msahn** [comment]
@lee1043 I am getting an error while testing the code. Do you have any idea on this error?

```
Traceback (most recent call last):
  File "/home/ahn6/PCMDI/pcmdi_metrics/branch/868_msa_precip_distribution_bimodality/pcmdi_metrics/pcmdi_metrics/precip_distribution/scripts_pcmdi/../precip_distribution_driver.py", line 10, in <module>
    from pcmdi_metrics.mean_climate.lib.pmp_parser import PMPParser
  File "<frozen zipimport>", line 259, in load_module
  File "/home/ahn6/anaconda3/envs/pcmdi_metrics_dev/lib/python3.9/site-packages/pcmdi_metrics-3.0.2-py3.9.egg/pcmdi_metrics/__init__.py", line 17, in <module>
  File "<frozen zipimport>", line 259, in load_module
  File "/home/ahn6/anaconda3/envs/pcmdi_metrics_dev/lib/python3.9/site-packages/pcmdi_metrics-3.0.2-py3.9.egg/pcmdi_metrics/io/__init__.py", line 2, in <module>
  File "<frozen zipimport>", line 259, in load_module
  File "/home/ahn6/anaconda3/envs/pcmdi_metrics_dev/lib/python3.9/site-packages/pcmdi_metrics-3.0.2-py3.9.egg/pcmdi_metrics/io/xcdat_openxml.py", line 6, in <module>
ModuleNotFoundError: No module named 'xmltodict'
```

https://github.com/PCMDI/pcmdi_metrics/blob/f0f6da4fd58bb4fada264c5529e0dcd78b0052dd/pcmdi_metrics/precip_distribution/precip_distribution_driver.py#L10

**Comment by lee1043** [comment]
@msahn `xmltodict` is [newly added to dependency](https://github.com/PCMDI/pcmdi_metrics/blob/fb960edabe0d429e99afa76783570e7012c7628f/conda-env/dev.yml#L23) recently and I guess it is not in your env. You can simply [install](https://anaconda.org/conda-forge/xmltodict) it in your conda env:
```
 conda install -c conda-forge xmltodict
```

**Comment by msahn** [comment]
> @msahn `xmltodict` is [newly added to dependency](https://github.com/PCMDI/pcmdi_metrics/blob/fb960edabe0d429e99afa76783570e7012c7628f/conda-env/dev.yml#L23) recently and I guess it is not in your env. You can simply [install](https://anaconda.org/conda-forge/xmltodict) it in your conda env:
> 
> ```
>  conda install -c conda-forge xmltodict
> ```

Thank you! It was resolved by installing `xmltodict`.
I have completed the code test, and it seems there are no more issues. I think this branch is ready to be merged into the main.

**Comment by lee1043** [comment]
I was able to reproduce the output json files and confirmed they are consistent.

**Review by lee1043 — APPROVED** [review]

File: pcmdi_metrics/precip_distribution/precip_distribution_driver.py [hunk 1]
```diff
@@ -51,54 +54,56 @@
...
-        data.append(model + "." + ens)
-print("Number of datasets:", len(file_list))
-print("Dataset:", data)
-
 # Read data -> Regrid -> Calculate metrics
 # It is working for daily average precipitation, in units of mm/day, with dimensions of (time,lat,lon)
+file_list = sorted(glob.glob(os.path.join(modpath, mod)))
+print(file_list)
+# f = xcdat.open_mfdataset(file_list)
+f = xr.open_mfdataset(file_list)
```
[the remark this answers is not in the document]
Comment (msahn) [reply pcmdi_metrics/precip_distribution/precip_distribution_driver.py:65]: When I tried to use `xcdat.open` for MERRA data, I got the below error:
```
>>> xcdat.open_mfdataset('/work/ahn6/obs/MERRA/pr/pr_day_reanalysis_MERRA_*.nc')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ahn6/anaconda3/envs/pcmdi_metrics_dev/lib/python3.9/site-packages/xcdat/dataset.py", line 223, in open_mfdataset
    ds = xr.open_mfdataset(
  File "/home/ahn6/anaconda3/envs/pcmdi_metrics_dev/lib/python3.9/site-packages/xarray/backends/api.py", line 1000, in open_mfdataset
    combined = combine_by_coords(
  File "/home/ahn6/anaconda3/envs/pcmdi_metrics_dev/lib/python3.9/site-packages/xarray/core/combine.py", line 982, in combine_by_coords
    concatenated = _combine_single_variable_hypercube(
  File "/home/ahn6/anaconda3/envs/pcmdi_metrics_dev/lib/python3.9/site-packages/xarray/core/combine.py", line 640, in _combine_single_variable_hypercube
    concatenated = _combine_nd(
  File "/home/ahn6/anaconda3/envs/pcmdi_metrics_dev/lib/python3.9/site-packages/xarray/core/combine.py", line 239, in _combine_nd
    combined_ids = _combine_all_along_first_dim(
  File "/home/ahn6/anaconda3/envs/pcmdi_metrics_dev/lib/python3.9/site-packages/xarray/core/combine.py", line 275, in _combine_all_along_first_dim
    new_combined_ids[new_id] = _combine_1d(
  File "/home/ahn6/anaconda3/envs/pcmdi_metrics_dev/lib/python3.9/site-packages/xarray/core/combine.py", line 298, in _combine_1d
    combined = concat(
  File "/home/ahn6/anaconda3/envs/pcmdi_metrics_dev/lib/python3.9/site-packages/xarray/core/concat.py", line 243, in concat
    return _dataset_concat(
  File "/home/ahn6/anaconda3/envs/pcmdi_metrics_dev/lib/python3.9/site-packages/xarray/core/concat.py", line 504, in _dataset_concat
    merged_vars, merged_indexes = merge_collected(
  File "/home/ahn6/anaconda3/envs/pcmdi_metrics_dev/lib/python3.9/site-packages/xarray/core/merge.py", line 302, in merge_collected
    merged_vars[name] = unique_variable(
  File "/home/ahn6/anaconda3/envs/pcmdi_metrics_dev/lib/python3.9/site-packages/xarray/core/merge.py", line 156, in unique_variable
    raise MergeError(
xarray.core.merge.MergeError: conflicting values for variable 'lon_bnds' on objects to be combined. You can skip this check by specifying compat='override'.
```

If I use `xarray.open`, the error does not occur.
```
>>> xr.open_mfdataset('/work/ahn6/obs/MERRA/pr/pr_day_reanalysis_MERRA_*.nc')
<xarray.Dataset>
Dimensions:    (lat: 361, lon: 952, time: 13574, bnds: 2)
Coordinates:
  * lat        (lat) float64 -90.0 -89.5 -89.0 -88.5 ... 88.5 89.0 89.5 90.0
  * lon        (lon) float64 0.0 0.6667 0.6667 1.333 ... 358.7 358.7 359.3 359.3
  * time       (time) datetime64[ns] 1979-01-01T12:00:00 ... 2016-02-29T12:00:00
Dimensions without coordinates: bnds
Data variables:
    time_bnds  (lon, time, bnds) datetime64[ns] dask.array<chunksize=(952, 365, 2), meta=np.ndarray>
    lat_bnds   (lon, time, lat, bnds) float64 dask.array<chunksize=(952, 365, 361, 2), meta=np.ndarray>
    lon_bnds   (time, lon, bnds) float64 dask.array<chunksize=(365, 952, 2), meta=np.ndarray>
    pr         (time, lat, lon) float32 dask.array<chunksize=(365, 361, 952), meta=np.ndarray>
Attributes: (12/20)
    institution:     Global Modeling and Assimilation Office, NASA Goddard Sp...
    institute_id:    NASA-GMAO
    experiment_id:   MERRA
    source:          MERRA Monthly 0.25x0.25 degree merged
    model_id:        GEOS-5
    contact:         MERRA, Steven Pawson (steven.pawson-1@nasa.gov)
    ...              ...
    Conventions:     CF-1.4
    project_id:      CREATE-IP
    table_id:        Table day (17 July 2013) 7d9fb6bca86b3be6e80f4fe674d87427
    title:           Reanalysis output prepared for CREATE-IP.
    modeling_realm:  atmos
    cmor_version:    2.9.1
```

[the remark this answers is not in the document]
Comment (lee1043) [reply pcmdi_metrics/precip_distribution/precip_distribution_driver.py:65]: Thank you for the comment. I brought this to the xcdat team: https://github.com/xCDAT/xcdat/issues/451

**Pull request merged**